File: Visual CSharp.NET/Forms Applications/Demo/frmSmtp.cs
//-----------------------------------------------------------------------
// <copyright file="frmSmtp.cs" company="Auron">
//     Copyright (c) Auron Software - www.auronsoftware.com
// </copyright>
// <author>Dennis van de Giessen</author>
//-----------------------------------------------------------------------
namespace Demo
{
  using System;
  using System.Windows.Forms;

  public partial class frmSmtp : Form
  {
    private const int IDX_NOTSECURE = 0;
    private const int IDX_SSLTLS = 1;
    private const int IDX_STARTTLS = 2;
    private const int IDX_XOAUTH2 = 3;

    private AxEmail.OAuth2 objOAuth2 = new AxEmail.OAuth2();
    private AxEmail.Message objMail = new AxEmail.Message();
    private AxEmail.Smtp objSmtp = new AxEmail.Smtp();
    private AxEmail.Constants objConstants = new AxEmail.Constants();
    
    private string sBearerToken = "";
    private DateTime dtBearerOrigin = DateTime.Now;

    private frmOAuth2Settings objFrmOauth2Settings = new frmOAuth2Settings();

    public frmSmtp()
    {
      InitializeComponent();

      txtHost.Text = "smtp.gmail.com";
    }

    private void frmSmtp_Load(object sender, EventArgs e)
    {
      // A license key is required to unlock this component after the trial period has expired.
      // Assign the LicenseKey property every time a new instance of this component is created (see code below). 
      // Alternatively, the LicenseKey property can be set automatically. This requires the license key to be stored in the registry.
      // For details, see manual, chapter "Product Activation".
      /*
      objSmtpServer.LicenseKey = "XXXXX-XXXXX-XXXXX";
      */

      if (txtPort.Text == string.Empty)
      {
        txtPort.Text = "25";
      }

      //Display Auron Email Component Info
      lblInfo.Text = String.Format("Auron Email Component {0}; Build: {1}; Module: {2}", objSmtp.Version, objSmtp.Build, objSmtp.Module);

      cbxSecurity.Items.Clear();
      cbxSecurity.Items.Add("Not Secure");
      cbxSecurity.Items.Add("SSL/TLS");
      cbxSecurity.Items.Add("STARTTLS");
      cbxSecurity.Items.Add("XOAUTH2");
      cbxSecurity.SelectedIndex = IDX_NOTSECURE;

      cbxFormat.Items.Clear();
      cbxFormat.Items.Add("Plain Text");
      cbxFormat.Items.Add("HTML");
      cbxFormat.SelectedIndex = 0;  // Plain Text

      cbxPriority.Items.Clear();
      cbxPriority.Items.Add("Highest Priority");
      cbxPriority.Items.Add("High Priority");
      cbxPriority.Items.Add("Normal Priority");
      cbxPriority.Items.Add("Low Priority");
      cbxPriority.Items.Add("Lowest Priority");
      cbxPriority.SelectedIndex = 2; // Normal Priority

      cbxEncoding.Items.Clear();
      cbxEncoding.Items.Add("Standard");    // Encoding
      cbxEncoding.Items.Add("UTF8");
      cbxEncoding.SelectedIndex = 0;

      txtResult.Text = string.Empty;
      txtLastSmtpResponse.Text = string.Empty;

      txtLogfile.Text = System.IO.Path.GetTempPath() + "Smtp.Log";

      UpdateControls();
    }

    private bool GetOAuth2BearerToken()
    {
      objOAuth2.LogFile = txtLogfile.Text;
      
      if (objFrmOauth2Settings.cbxFlow.SelectedIndex == frmOAuth2Settings.IDX_AUTHFLOW)
        objOAuth2.Flow = objConstants.OAUTH2_FLOW_AUTHCODE;
      else
        objOAuth2.Flow = objConstants.OAUTH2_FLOW_DEVICECODE;

      // set all oauth properties
      objOAuth2.ClientID = objFrmOauth2Settings.txtClientID.Text;
      objOAuth2.ClientSecret = objFrmOauth2Settings.txtClientSecret.Text;
      objOAuth2.Scope = objFrmOauth2Settings.txtScope.Text;
      objOAuth2.RedirectUrl = objFrmOauth2Settings.txtRedirectUrl.Text;
      objOAuth2.AuthCodeUrl = objFrmOauth2Settings.txtAuthCodeUrl.Text;
      objOAuth2.DeviceCodeUrl = objFrmOauth2Settings.txtDeviceCodeUrl.Text;
      objOAuth2.TokenExchangeUrl = objFrmOauth2Settings.txtTokenExchangeUrl.Text;

      objOAuth2.Login();
      SetResult(objOAuth2.LastError);

      int nTokenExchangeTimeoutMs = 0;
      if (objOAuth2.LastError != 0) return false;
      if (objOAuth2.Flow == objConstants.OAUTH2_FLOW_DEVICECODE)
      { // show device code!        
        frmShowDeviceCode objFrmShowCode = new frmShowDeviceCode();
        objFrmShowCode.lblUserCode.Text = objOAuth2.UserCode;
        objFrmShowCode.ShowDialog();

        nTokenExchangeTimeoutMs = objOAuth2.UserCodeExpInSeconds * 1000;
      } else
        nTokenExchangeTimeoutMs = 120000;

      objOAuth2.WaitForTokens(nTokenExchangeTimeoutMs);
      SetResult(objOAuth2.LastError);
      if (objOAuth2.LastError == 0)
      {
        dtBearerOrigin = DateTime.Now;
        sBearerToken = objOAuth2.BearerToken;
      }

      return objOAuth2.LastError == 0;
    }

    private bool RefreshOAuth2BearerToken()
    {
      objOAuth2.RefreshBearerToken();
      SetResult(objOAuth2.LastError);
      return objOAuth2.LastError == 0;
    }

    private void btnSend_Click(object sender, EventArgs e)
    {
      Cursor CurCursor = Cursor;
      Cursor = Cursors.WaitCursor;

      txtResult.Text = txtLastSmtpResponse.Text = "";

      // Mail: Clear (good pratise)
      objMail.Clear();

      // Mail: From
      objMail.FromName = txtFromName.Text;
      objMail.FromAddress = txtFromAddress.Text;

      // Mail: Subject
      objMail.Subject = txtSubject.Text;

      // Mail: Priority
      objMail.Priority = cbxPriority.SelectedIndex + 1;

      // Mail: Encoding
      objMail.Encoding = cbxEncoding.SelectedIndex == 1 ? objConstants.EMAIL_MESSAGE_ENCODING_UTF8 : objConstants.EMAIL_MESSAGE_ENCODING_DEFAULT;

      // Mail: Body
      if (cbxFormat.SelectedIndex == 0)
        objMail.BodyPlainText = txtBody.Text;
      else
        objMail.BodyHtml = txtBody.Text;


      // Mail: TO recipient(s)  
      if (txtToAddress.Text != string.Empty)
        objMail.AddTo(txtToAddress.Text, txtToAddress.Text);

      // Mail: CC recipient(s)  
      if (objMail.LastError == 0 && txtCcAddress.Text != string.Empty)
        objMail.AddCc(txtCcAddress.Text, txtCcAddress.Text);

      // Mail: BCC recipient(s)
      if (objMail.LastError == 0 && txtBccAddress.Text != string.Empty)
        objMail.AddBcc(txtBccAddress.Text, txtBccAddress.Text);

      // Mail: Attachment(s)
      if (objMail.LastError == 0 && txtAttachments.Text != string.Empty)
        objMail.AddAttachment(txtAttachments.Text);

      // Mail: If a function failed then quit
      if (objMail.LastError != 0)
      {
        txtResult.Text = objMail.LastError + ": " + objMail.GetErrorDescription(objMail.LastError);
        Cursor = CurCursor;
        return;
      }

      // Smtp: Clear (good practise)
      objSmtp.Clear();
      objSmtp.LogFile = txtLogfile.Text;

      // Smtp: Set Secure if secure communications is required 
      int nSecIdx = cbxSecurity.SelectedIndex;
      if (nSecIdx == IDX_SSLTLS)
          objSmtp.SetSecure(Int32.Parse(txtPort.Text));
      else
        objSmtp.HostPort = Int32.Parse(txtPort.Text);
      
      // only for OAuth authenticiation
      if (nSecIdx == IDX_XOAUTH2)
      { // for OAuth2 we need to have a bearer token so try to obtain one
        if (sBearerToken == "")
        { 
          if (!GetOAuth2BearerToken()) 
            return;          
        }        

        // if the bearer token has expired then refresh it
        TimeSpan tsSpan = DateTime.Now - dtBearerOrigin;
        if (tsSpan.TotalSeconds >= objOAuth2.BearerExpInSeconds)
        {
          if (!RefreshOAuth2BearerToken())
            return;
        }

        // set the resulting bearer token in the SMTP object
        objSmtp.BearerToken = sBearerToken;
      }

      // Smtp: Connect
      objSmtp.Connect(txtHost.Text, txtAccount.Text, txtPassword.Text);

      // Smtp: Send
      if (objSmtp.LastError == 0)
      {
        // Sends out the E-Mail
        objSmtp.Send(objMail);
      }

      SetResult(objSmtp.LastError);

      // Disconnects the SMTP server
      objSmtp.Disconnect();

      Cursor = CurCursor;
    }

    private void SetResult(int nResult)
    {
      txtResult.Text = nResult + ": " + objSmtp.GetErrorDescription(nResult);
      txtLastSmtpResponse.Text = objSmtp.LastSmtpResponse;
    }

    private void btnBrowseAttachment_Click(object sender, EventArgs e)
    {
      System.Windows.Forms.OpenFileDialog fdlg;

      fdlg = new OpenFileDialog();
      fdlg.Title = "Select Attachment";
      fdlg.InitialDirectory = "C:\\";
      fdlg.Filter = "All files (*.*)|*.*";
      fdlg.FilterIndex = 1;
      fdlg.RestoreDirectory = true;
      fdlg.Multiselect = true;
      if (fdlg.ShowDialog() != DialogResult.OK)
        return;

      foreach (string strFile in fdlg.FileNames)
      {
        if (txtAttachments.Text != string.Empty)
          txtAttachments.Text += "|";
        txtAttachments.Text += strFile;
      }
    }

    private void btnViewLog_Click(object sender, EventArgs e)
    {
      if (System.IO.File.Exists(txtLogfile.Text))
      {
        System.Diagnostics.Process.Start(txtLogfile.Text);
      }
    }

    private void cbAuthentication_CheckedChanged(object sender, EventArgs e)
    {
      UpdateControls();
    }

    private void UpdateControls()
    {
      int nIdx = cbxSecurity.SelectedIndex;
      btnOAuth.Enabled = nIdx == IDX_XOAUTH2;
    }

    private void cbxSecurity_SelectedIndexChanged(object sender, EventArgs e)
    {
      int nIdx = cbxSecurity.SelectedIndex;

      if (nIdx == IDX_NOTSECURE)
        txtPort.Text = "25";
      else if (nIdx == IDX_SSLTLS)
        txtPort.Text = "465";
      else if (nIdx == IDX_STARTTLS)
        txtPort.Text = "587";
      else
        txtPort.Text = "587";

      UpdateControls();
    }

    private void btnOAuth_Click(object sender, EventArgs e)
    {
      objFrmOauth2Settings.ShowDialog();
      sBearerToken = ""; // reset bearer token cache
    }

    private void txtHost_TextChanged(object sender, EventArgs e)
    {
      sBearerToken = ""; // reset bearer token cache
    }
  }
}

File: Visual Basic.NET/Forms Applications/Demo/frmSmtp.vb
'-----------------------------------------------------------------------
' <copyright file="frmSmtp.vb" company="Auron">
'     Copyright (c) Auron Software - www.auronsoftware.com
' </copyright>
' <author>Dennis van de Giessen</author>
'-----------------------------------------------------------------------
Imports AxEmail

Public Class frmSmtp
  Private Const IDX_NOTSECURE As Integer = 0
  Private Const IDX_SSLTLS As Integer = 1
  Private Const IDX_STARTTLS As Integer = 2
  Private Const IDX_XOAUTH2 As Integer = 3

  Private objMail As AxEmail.Message = New AxEmail.Message()
  Private objSmtpServer As AxEmail.Smtp = New AxEmail.Smtp()
  Private objConstants As AxEmail.Constants = New AxEmail.Constants()

  Private objOAuth2 As AxEmail.OAuth2 = New AxEmail.OAuth2()
  Private sBearerToken As String = ""
  Private dtBearerOrigin As DateTime = DateTime.Now
  Private objFrmOauth2Settings As frmOAuth2Settings = New frmOAuth2Settings()

  Private Sub frmSmtp_Load(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Load

    ' A license key is required to unlock this component after the trial period has expired.
    ' Assign the LicenseKey property every time a new instance of this component is created (see code below). 
    ' Alternatively, the LicenseKey property can be set automatically. This requires the license key to be stored in the registry.
    ' For details, see manual, chapter "Product Activation".
    '
    ' objSmtpServer.LicenseKey = "XXXXX-XXXXX-XXXXX"

    If (txtPort.Text = String.Empty) Then
      txtPort.Text = "25"
    End If

    txtHost.Text = "smtp.gmail.com"

    'Display Auron Email Component Info
    lblInfo.Text = String.Format("Auron Email Component {0}; Build: {1}; Module: {2}", objSmtpServer.Version, objSmtpServer.Build, objSmtpServer.Module)

    cbxSecurity.Items.Clear()
    cbxSecurity.Items.Add("Not Secure")
    cbxSecurity.Items.Add("SSL/TLS")
    cbxSecurity.Items.Add("STARTTLS")
    cbxSecurity.Items.Add("XOAUTH2")
    cbxSecurity.SelectedIndex = IDX_NOTSECURE

    cbxFormat.Items.Clear()
    cbxFormat.Items.Add("Plain Text")
    cbxFormat.Items.Add("HTML")
    cbxFormat.SelectedIndex = 0  ' Plain Text

    cbxPriority.Items.Clear()
    cbxPriority.Items.Add("Highest Priority")
    cbxPriority.Items.Add("High Priority")
    cbxPriority.Items.Add("Normal Priority")
    cbxPriority.Items.Add("Low Priority")
    cbxPriority.Items.Add("Lowest Priority")
    cbxPriority.SelectedIndex = 2 '  Normal Priority

    cbxEncoding.Items.Clear()
    cbxEncoding.Items.Add("Standard") ' Encoding
    cbxEncoding.Items.Add("UTF8")
    cbxEncoding.SelectedIndex = 0

    txtResult.Text = String.Empty
    txtLastSmtpResponse.Text = String.Empty

    txtLogfile.Text = System.IO.Path.GetTempPath() + "Smtp.log"

    UpdateControls()
  End Sub

  Private Sub SetResult(ByVal nResult As Integer)
    txtResult.Text = nResult & ": " & objSmtpServer.GetErrorDescription(nResult)
    txtLastSmtpResponse.Text = objSmtpServer.LastSmtpResponse
  End Sub

  Private Function GetOAuth2BearerToken() As Boolean
    objOAuth2.LogFile = txtLogfile.Text

    If objFrmOauth2Settings.cbxFlow.SelectedIndex = frmOAuth2Settings.IDX_AUTHFLOW Then
      objOAuth2.Flow = objConstants.OAUTH2_FLOW_AUTHCODE
    Else
      objOAuth2.Flow = objConstants.OAUTH2_FLOW_DEVICECODE
    End If

    objOAuth2.ClientID = objFrmOauth2Settings.txtClientID.Text
    objOAuth2.ClientSecret = objFrmOauth2Settings.txtClientSecret.Text
    objOAuth2.Scope = objFrmOauth2Settings.txtScope.Text
    objOAuth2.RedirectUrl = objFrmOauth2Settings.txtRedirectUrl.Text
    objOAuth2.AuthCodeUrl = objFrmOauth2Settings.txtAuthCodeUrl.Text
    objOAuth2.DeviceCodeUrl = objFrmOauth2Settings.txtDeviceCodeUrl.Text
    objOAuth2.TokenExchangeUrl = objFrmOauth2Settings.txtTokenExchangeUrl.Text
    objOAuth2.Login()
    SetResult(objOAuth2.LastError)
    Dim nTokenExchangeTimeoutMs As Integer = 0
    If objOAuth2.LastError <> 0 Then Return False

    If objOAuth2.Flow = objConstants.OAUTH2_FLOW_DEVICECODE Then
      Dim objFrmShowCode As frmShowDeviceCode = New frmShowDeviceCode()
      objFrmShowCode.lblUserCode.Text = objOAuth2.UserCode
      objFrmShowCode.ShowDialog()
      nTokenExchangeTimeoutMs = objOAuth2.UserCodeExpInSeconds * 1000
    Else
      nTokenExchangeTimeoutMs = 120000
    End If

    objOAuth2.WaitForTokens(nTokenExchangeTimeoutMs)
    SetResult(objOAuth2.LastError)

    If objOAuth2.LastError = 0 Then
      dtBearerOrigin = DateTime.Now
      sBearerToken = objOAuth2.BearerToken
    End If

    Return objOAuth2.LastError = 0
  End Function

  Private Function RefreshOAuth2BearerToken() As Boolean
    objOAuth2.RefreshBearerToken()
    SetResult(objOAuth2.LastError)
    If objOAuth2.LastError = 0 Then
      dtBearerOrigin = DateTime.Now
      sBearerToken = objOAuth2.BearerToken
    End If
    Return objOAuth2.LastError = 0
  End Function

  Sub UpdateControls()
    Dim nIdx As Integer = cbxSecurity.SelectedIndex
    btnOAuth.Enabled = nIdx = IDX_XOAUTH2
  End Sub

  Private Sub btnSend_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnSend.Click

    Dim CurCursor = Cursor.Current

    Cursor = Cursors.WaitCursor

    txtResult.Text = ""
    txtLastSmtpResponse.Text = ""

    ' Mail: Clear (good pratise)
    objMail.Clear()

    ' Mail: From
    objMail.FromName = txtFromName.Text
    objMail.FromAddress = txtFromAddress.Text

    ' Mail: Subject
    objMail.Subject = txtSubject.Text

    ' Mail: Priority
    objMail.Priority = cbxPriority.SelectedIndex + 1

    ' Mail: Encoding
    If (cbxEncoding.SelectedIndex = 0) Then
      objMail.Encoding = objConstants.EMAIL_MESSAGE_ENCODING_DEFAULT
    Else
      objMail.Encoding = objConstants.EMAIL_MESSAGE_ENCODING_UTF8
    End If

    ' Mail: Body
    If (cbxFormat.SelectedIndex = 0) Then
      objMail.BodyPlainText = txtBody.Text
    Else
      objMail.BodyHtml = txtBody.Text
    End If

    ' Mail: TO recipient(s)
    If (txtToAddress.Text <> String.Empty) Then
      objMail.AddTo(txtToAddress.Text, txtToAddress.Text)
    End If

    ' Mail: CC recipient(s)
    If (objMail.LastError = 0 And txtCcAddress.Text <> String.Empty) Then
      objMail.AddCc(txtCcAddress.Text, txtCcAddress.Text)
    End If

    ' Mail: BCC recipient(s)
    If (objMail.LastError = 0 And txtBccAddress.Text <> String.Empty) Then
      objMail.AddBcc(txtBccAddress.Text, txtBccAddress.Text)
    End If

    ' Mail: Attachments
    If (objMail.LastError = 0 And txtBccAddress.Text <> String.Empty) Then
      objMail.AddBcc(txtBccAddress.Text, txtBccAddress.Text)
    End If

    ' Mail: Add attachment(s)
    If (txtAttachments.Text <> String.Empty) Then
      objMail.AddAttachment(txtAttachments.Text)
    End If


    ' Mail: If a function failed then quit
    If (objMail.LastError <> 0) Then
      txtResult.Text = objMail.LastError.ToString() & ": " & objMail.GetErrorDescription(objMail.LastError)
      Exit Sub
    End If

    ' Smtp: Clear (good practise)        
    objSmtpServer.Clear()
    objSmtpServer.LogFile = txtLogfile.Text

    Dim nSecIdx As Integer = cbxSecurity.SelectedIndex

    If nSecIdx = IDX_SSLTLS Then
      objSmtpServer.SetSecure(Int32.Parse(txtPort.Text))
    Else
      objSmtpServer.HostPort = Int32.Parse(txtPort.Text)
    End If

    If nSecIdx = IDX_XOAUTH2 Then
      If sBearerToken = "" Then
        If Not GetOAuth2BearerToken() Then Return
      End If

      Dim tsSpan As TimeSpan = DateTime.Now - dtBearerOrigin

      If tsSpan.TotalSeconds >= objOAuth2.BearerExpInSeconds Then
        If Not RefreshOAuth2BearerToken() Then Return
      End If

      objSmtpServer.BearerToken = sBearerToken
    End If

    ' Smtp: Connect
    objSmtpServer.Connect(txtHost.Text, txtAccount.Text, txtPassword.Text)

    ' Smtp: Send
    If (objSmtpServer.LastError = 0) Then
      objSmtpServer.Send(objMail)
    End If

    txtResult.Text = objSmtpServer.LastError.ToString() & ": " & objSmtpServer.GetErrorDescription(objSmtpServer.LastError)
    txtLastSmtpResponse.Text = objSmtpServer.LastSmtpResponse

    ' Smtp: Disconnect
    objSmtpServer.Disconnect()

    Cursor = CurCursor
  End Sub

  Private Sub btnBrowseAttachment_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnBrowseAttachment.Click
    Dim fdlg As System.Windows.Forms.OpenFileDialog = New System.Windows.Forms.OpenFileDialog

    fdlg.Title = "Select Attachment"
    fdlg.InitialDirectory = "C:\\"
    fdlg.Filter = "All files (*.*)|*.*"
    fdlg.FilterIndex = 1
    fdlg.RestoreDirectory = True
    fdlg.Multiselect = True
    If (fdlg.ShowDialog() <> DialogResult.OK) Then
      Return
    End If

    For Each strFile As String In fdlg.FileNames
      If (txtAttachments.Text <> String.Empty) Then
        txtAttachments.Text += "|"
      End If
      txtAttachments.Text += strFile
    Next
  End Sub

  Private Sub btnViewLog_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnViewLog.Click
    If (System.IO.File.Exists(txtLogfile.Text)) Then
      System.Diagnostics.Process.Start(txtLogfile.Text)
    End If
  End Sub

  Private Sub cbxSecurity_SelectedIndexChanged(sender As Object, e As EventArgs) Handles cbxSecurity.SelectedIndexChanged
    Dim nIdx As Integer = cbxSecurity.SelectedIndex

    If nIdx = IDX_NOTSECURE Then
      txtPort.Text = "25"
    ElseIf nIdx = IDX_SSLTLS Then
      txtPort.Text = "465"
    ElseIf nIdx = IDX_STARTTLS Then
      txtPort.Text = "587"
    Else
      txtPort.Text = "587"
    End If

    UpdateControls()
  End Sub

  Private Sub btnOAuth_Click(sender As Object, e As EventArgs) Handles btnOAuth.Click
    objFrmOauth2Settings.ShowDialog()
    sBearerToken = ""  ' reset bearer token cache
  End Sub

  Private Sub txtHost_TextChanged(sender As Object, e As EventArgs) Handles txtHost.TextChanged
    sBearerToken = ""  ' reset bearer token cache
  End Sub
End Class